icu4c/source/common/uloc.cpp

```diff
@@ -1184,104 +1184,128 @@ _getLanguage(const char* localeID,
-
     *pEnd = localeID;
 
-    /* copy the second item as far as possible and count its length */
-    while(!_isTerminator(localeID[idLen]) && !_isIDSeparator(localeID[idLen])
-            && uprv_isASCIILetter(localeID[idLen])) {
-        idLen++;
-    }
+    constexpr int32_t LENGTH = 4;
 
-    /* If it's exactly 4 characters long, then it's a script and not a country. */
-    if (idLen == 4) {
-        int32_t i;
-        *pEnd = localeID + idLen;
-        if (idLen >= 1) {
-            result.append((char)uprv_toupper(*(localeID++)), status);
-        }
-        for (i = 1; i < idLen; i++) {
-            result.append((char)uprv_tolower(*(localeID++)), status);
-        }
+    int32_t len = 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len]) &&
+            uprv_isASCIILetter(localeID[len])) {
+        len++;
     }
+    if (len != LENGTH) { return; }
+
+    *pEnd = localeID + LENGTH;
+    if (sink == nullptr) { return; }
 
-    if (sink != nullptr && !result.isEmpty()) {
-        sink->Append(result.data(), result.length());
+    char scratch[LENGTH];
+    int32_t capacity = 0;
+    char* buffer = sink->GetAppendBuffer(
+            LENGTH, LENGTH, scratch, UPRV_LENGTHOF(scratch), &capacity);
+    U_ASSERT(capacity >= LENGTH);
+
+    buffer[0] = uprv_toupper(localeID[0]);
+    for (int32_t i = 1; i < LENGTH; ++i) {
+        buffer[i] = uprv_tolower(localeID[i]);
     }
+
+    sink->Append(buffer, LENGTH);
 }
 
 static void
 _getRegion(const char* localeID,
            ByteSink* sink,
-           const char** pEnd,
-           UErrorCode& status) {
+           const char** pEnd) {
     U_ASSERT(pEnd != nullptr);
-    CharString result;
-    int32_t idLen=0;
+    *pEnd = localeID;
+
+    constexpr int32_t MINLEN = 2;
+    constexpr int32_t MAXLEN = ULOC_COUNTRY_CAPACITY - 1;  // Minus NUL.
 
-    /* copy the country as far as possible and count its length */
-    while(!_isTerminator(localeID[idLen]) && !_isIDSeparator(localeID[idLen])) {
-        result.append((char)uprv_toupper(localeID[idLen]), status);
-        idLen++;
+    int32_t len = 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len])) {
+        len++;
     }
+    if (len < MINLEN || len > MAXLEN) { return; }
 
-    /* the country should be either length 2 or 3 */
-    if (idLen == 2 || idLen == 3) {
-        /* convert 3 character code to 2 character code if possible *CWB*/
-        if(idLen==3) {
-            int32_t offset = _findIndex(COUNTRIES_3, result.data());
-            if(offset>=0) {
-                result.clear();
-                result.append(COUNTRIES[offset], status);
-            }
-        }
-        localeID+=idLen;
-    } else {
-        result.clear();
+    *pEnd = localeID + len;
+    if (sink == nullptr) { return; }
+
+    char scratch[ULOC_COUNTRY_CAPACITY];
+    int32_t capacity = 0;
+    char* buffer = sink->GetAppendBuffer(
+            ULOC_COUNTRY_CAPACITY,
+            ULOC_COUNTRY_CAPACITY,
```

Comment:
rather than capacity=ULOC_COUNTRY_CAPACITY, we should use min(len, 4) to maximize use of tight ByteSink buffers; with a comment that we need 4 for NUL-terminating a 3-letter code for lookup of a 2-letter code

Reply:
Done.